docs/integration-testing.md

```diff
@@ -19,53 +18,66 @@ typically run faster. Sometimes that's not (easily) possible, especially when an
-any residue before and after the test.
+We encourage you to use these available annotation but if your integration test does not fit in one of these available annotations and you want to categorize them base on their technologies then feel free to create a new annotations but make sure to use composite annotations which contains `@IntegrationTest`. If you do not wish to categorize base on their technologies then you can use already available `@IntegrationTest` annotation.
+
+- By default JUnit test runner ignores all integration tests because in root `build.gradle.kts` file we have excluded all tests with `integration-test` Junit tag.
+
+All integration tests should specify annotation to categorize them and the `"...IntegrationTest"` postfix to distinguish them clearly from unit tests. They should reside in the same package as unit tests because all tests should maintain package consistency to their test subject.
+
+Any credentials, secrets, passwords, etc. that are required by the integration tests should be passed in using environment variables. A good way to access them is `ConfigurationFunctions.propOrEnv()` because then the credentials can also be supplied via system properties.
+
+There is no one-size-fits-all guideline whether to perform setup tasks in the `@BeforeAll` or `@BeforeEach`, it will depend on the concrete system you're using. As a general rule of thumb long-running one-time setup should be done in the `@BeforeAll` so as not to extend the run-time of the test unnecessarily. In contrast, in most cases it is **not** advisable to deploy/provision the external system itself in either of those methods. In other words, provisioning a CosmosDB or spinning up a Postgres docker container directly from test code should generally be avoided, because it will introduce code that has nothing to do with the test and may cause security problems (privilege escalation through the Docker API), etc.
+
+This does not at all discourage the use of external test environments like containers, rather, the external system should be deployed in the CI script (e.g. through Github's `services` feature), or there might even be a dedicated test instance running continuously, e.g. a CosmosDB test instance in Azure. In the latter case we need to be careful to avoid conflicts (e.g. database names) when multiple test runners access that system simultaneously and to properly clean-up any residue before and after the test.
 
 ## Running them locally
 
-The JUnit runner won't pick up integration tests unless the `RUN_INTEGRATION_TEST` environment variable is set to `true`
-. Also, don't forget to define any credentials that are needed.
+As mentioned above the JUnit runner won't pick up integration tests unless a tag is provided. For instance if needed to run integration tests related to `Azure CosmosDB` then it can be achieved by passing the `IncludeTags` parameter to the `gradlew` command:
+
+```bash
+./gradlew test -DIncludeTags="azure-cosmos-db-integration-test"
```

Comment:
```suggestion
./gradlew test -p path/to/module -DIncludeTags="azure-cosmos-db-integration-test"
```

Probably more realistic, since a dev is unlikely to have all emulators running